flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/topic/TopicMetadataListener.java

```diff
@@ -0,0 +1,172 @@
+import static org.apache.flink.connector.pulsar.common.utils.PulsarExceptionUtils.sneakyAdmin;
+import static org.apache.flink.connector.pulsar.source.enumerator.topic.TopicNameUtils.isPartitioned;
+import static org.apache.flink.connector.pulsar.source.enumerator.topic.TopicNameUtils.topicNameWithPartition;
+
+/**
+ * We need the latest topic metadata for making sure the newly created topic partitions would be
+ * used by Pulsar sink. This routing policy would be different comparing with Pulsar Client built-in
+ * logic. We use flink's ProcessingTimer as the executor.
+ */
+@Internal
+public class TopicMetadataListener implements Serializable, Closeable {
+    private static final long serialVersionUID = 6186948471557507522L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(TopicMetadataListener.class);
+
+    private final List<String> partitionedTopics;
+    private final Map<String, Integer> topicMetadata;
+    private volatile ImmutableList<String> availableTopics;
+
+    // Dynamic fields.
+    private transient PulsarAdmin pulsarAdmin;
+    private transient Long topicMetadataRefreshInterval;
+    private transient ProcessingTimeService timeService;
+
+    public TopicMetadataListener() {
+        this(emptyList());
+    }
+
+    public TopicMetadataListener(List<String> topics) {
+        this.partitionedTopics = new ArrayList<>(topics.size());
+        this.topicMetadata = new HashMap<>(topics.size());
+        this.availableTopics = ImmutableList.of();
+
+        for (String topic : topics) {
+            if (isPartitioned(topic)) {
+                partitionedTopics.add(topic);
+            } else {
+                // This would be updated when open writing.
+                topicMetadata.put(topic, -1);
+            }
+        }
+    }
+
+    /** Register the topic metadata update in process time service. */
+    public void open(SinkConfiguration sinkConfiguration, ProcessingTimeService timeService) {
+        if (topicMetadata.isEmpty()) {
+            LOG.info("No topics have been provided, skip listener initialize.");
+            return;
+        }
+
+        // Initialize listener properties.
+        this.pulsarAdmin = createAdmin(sinkConfiguration);
+        this.topicMetadataRefreshInterval = sinkConfiguration.getTopicMetadataRefreshInterval();
+        this.timeService = timeService;
+
+        // Initialize the topic metadata. Quit if fail to connect to Pulsar.
+        sneakyAdmin(this::updateTopicMetadata);
+
+        // Register time service.
+        triggerNextTopicMetadataUpdate(true);
+    }
+
+    /**
+     * Return all the available topic partitions. We would recalculate the partitions if the topic
+     * metadata hsa been changed. Otherwise, we would return the cached result for better
+     * performance.
+     */
+    public List<String> availableTopics() {
+        if (availableTopics.isEmpty()
+                && (!partitionedTopics.isEmpty() || !topicMetadata.isEmpty())) {
```

Comment:
Just out of curiousity, we evaluate availableTopics when we call it (evaluate: retrieving data from topicMetadata) in this method instead of in updateTopicMetadata(), is there any special reason to do so ?

Reply:
I should clarify that you are wrong. We only evaluate the `availableTopics` when it needs updating. This is a cached result, which should be created every time we call this method.

Putting the evaluate logic here is just a design flavor. XD